src/dataflow/src/decode/mod.rs

```diff
@@ -220,97 +211,18 @@ where
     fn log_error_count(&mut self) {}
 }
 
-/// Inner method for decoding an upsert source
-/// Mostly, this inner method exists that way static dispatching
-/// can be used for different combinations of key-value decoders
-/// as opposed to dynamic dispatching
-fn decode_upsert_inner<G, K, V>(
-    stream: &Stream<G, ((Vec<u8>, SourceData), Timestamp)>,
-    mut key_decoder_state: K,
-    mut value_decoder_state: V,
-    op_name: &str,
-) -> Stream<G, (Row, Option<Row>, Timestamp)>
-where
-    G: Scope<Timestamp = Timestamp>,
-    K: DecoderState + 'static,
-    V: DecoderState + 'static,
-{
-    stream.unary(
-        Exchange::new(|x: &((Vec<u8>, _), _)| (x.0).hashed()),
-        &op_name,
-        move |_, _| {
-            move |input, output| {
-                input.for_each(|cap, data| {
-                    let mut session = output.session(&cap);
-                    for ((key, data), time) in data.iter() {
-                        if key.is_empty() {
-                            error!("{}", "Encountered empty key");
-                            continue;
-                        }
-                        match key_decoder_state.decode_key(key) {
-                            Ok(key) => {
-                                if data.value.is_empty() {
-                                    session.give((key, None, *cap.time()));
-                                } else {
-                                    value_decoder_state.give_key_value(
-                                        key,
-                                        &data.value,
-                                        data.position,
-                                        data.upstream_time_millis,
-                                        &mut session,
-                                        *time,
-                                    );
-                                }
-                            }
-                            Err(err) => {
-                                error!("{}", err);
-                            }
-                        }
-                    }
-                });
-                key_decoder_state.log_error_count();
-                value_decoder_state.log_error_count();
-            }
-        },
-    )
-}
-
-pub(crate) fn decode_upsert<G>(
-    stream: &Stream<G, ((Vec<u8>, SourceData), Timestamp)>,
-    value_encoding: DataEncoding,
-    key_encoding: DataEncoding,
+/// Get the `DecoderState` corresponding to a particular `DataEncoding`.
+/// Note that for code simplicity, this uses dynamic dispatch, which is said to
+/// be slower than using templates. The use of dynamic dispatch has not been
+/// observed to noticeably impact upsert performance at the time of this writing.
```

Comment:
No urgency for this PR, but we can talk through how to reshape this to use generics if that is appealing. It looks like `DataEncoding` might want to be a trait, with an associated type `DecoderState`.

Reply:
I would love to be a fly on the wall for that. I've found it really challenging to reason about generics

Reply:
Will add as a TODO